tests/p2p/test_sync_rate_limiter.py

```diff
@@ -40,6 +42,136 @@ def test_sync_rate_limiter(self):
+
+        self.simulator.run(30)
+
+        # Send tips should not be called any further since the connection has already been closed.
+        self.assertEqual(sync1._send_tips.call_count, 1)
+        # Residual delayed calls
+        self.assertEqual(len(sync1._send_tips_call_later), 2)
+        # The residual delayed calls should have been canceled
+        for call_later in sync1._send_tips_call_later:
+            self.assertFalse(call_later.active())
+
+    def test_sync_rate_limiter_delayed_calls_draining(self):
+        # Test the draining of delayed calls from _send_tips_call_later list
+        manager1 = self.create_peer()
+        manager2 = self.create_peer()
+        manager2.connections.MAX_ENABLED_SYNC = 0
+
+        conn12 = FakeConnection(manager1, manager2, latency=0.05)
+        self.simulator.add_connection(conn12)
+        self.simulator.run(3600)
+
+        connections = manager2.connections
+        connections.rate_limiter.reset(connections.GlobalRateLimiter.SEND_TIPS)
+        connections.enable_rate_limiter(1, 1)
+
+        connected_peers2 = list(manager2.connections.connected_peers.values())
+        self.assertEqual(1, len(connected_peers2))
+
+        protocol1 = connected_peers2[0]
+        sync1 = protocol1.state.sync_manager
+
+        sync1.send_tips()
+        self.assertEqual(len(sync1._send_tips_call_later), 0)
+
+        sync1.send_tips()
+        self.assertEqual(len(sync1._send_tips_call_later), 1)
+
+        sync1.send_tips()
+        self.assertEqual(len(sync1._send_tips_call_later), 2)
+
+        sync1.send_tips()
+        self.assertEqual(len(sync1._send_tips_call_later), 3)
+
+        self.simulator.run(30)
+
+        # Without disconnection, all the delayed calls
+        # should have been executed
+        self.assertEqual(len(sync1._send_tips_call_later), 0)
+
+    def test_sync_rate_limiter_delayed_calls_stop(self):
+        # Test the draining of delayed calls from _send_tips_call_later list
+        manager1 = self.create_peer()
+        manager2 = self.create_peer()
+        manager2.connections.MAX_ENABLED_SYNC = 0
+
+        conn12 = FakeConnection(manager1, manager2, latency=0.05)
+        self.simulator.add_connection(conn12)
+        self.simulator.run(3600)
+
+        connections = manager2.connections
+        connections.rate_limiter.reset(connections.GlobalRateLimiter.SEND_TIPS)
+        connections.enable_rate_limiter(1, 1)
+
+        connected_peers2 = list(manager2.connections.connected_peers.values())
+        self.assertEqual(1, len(connected_peers2))
+
+        protocol1 = connected_peers2[0]
+        sync1 = protocol1.state.sync_manager
+
+        sync1.send_tips()
+        self.assertEqual(len(sync1._send_tips_call_later), 0)
+
+        from hathor.conf import HathorSettings
+        settings = HathorSettings()
+
+        max_delayed_calls = settings.MAX_GET_TIPS_DELAYED_CALLS
+        for count in range(max_delayed_calls + 1):
```

Comment:
Run at the limit `max_delayed_calls` and confirm that the connection is not closed. Then, call `send_tips()` and check that the connection was closed.

Reply:
Done at 47a4acb2fb7932a1007d080a7f7e935f5746d208

Reply:
Thanks!